Add ability for Composer to save email to draft

yifanplanet opened the pull request:
# Code changes

- Added ability for composer to save email as a new draft
- Composer can update existing draft if draft message id is provided
- Uploaded files id is saved along with draft message, so saved draft will have the same file attached to it.


# Screenshot

![Screen Shot 2022-02-09 at 12 44 43 AM](https://user-images.githubusercontent.com/14408339/153129508-652b9355-e884-41ec-90dc-6709932c60c1.png)

# Readiness checklist

- [x] Added changes to component `CHANGELOG.md`
- [x] Cypress tests passing?
- [x] New cypress tests added?
- [x] Included before/after screenshots, if the change is visual

# License

I confirm that this contribution is made under the terms of the MIT license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

yifanplanet commented:
Currently individual draft message is not showing in expanded thread in Email component, in order to better integrate this, created the feature story:
https://app.shortcut.com/nylas/story/81020/display-and-edit-draft-as-individual-message-in-a-email-component-thread

philrenaud reviewed (changes requested):
Code and tests look good. A couple change requests:
- the Save as Drafts button should be added to the tabindex / be tabbable. Right now after tabbing out of the body, you go right to the Send button, which is good and we shouldn't change. But after that, we should tab to the Save as Drafts button
- the existence of this button should be predicated on a `show_save_as_draft` prop, or something similarly named. Don't assume end users will always want this. It's fine for it to default to true.

pooja169usp reviewed (changes requested):
@yifanplanet A couple things I noticed:
1. When a message has an attachedFile and is saved as a draft, we should make an additional request to `/files` endpoint, get the file object and add it to the draft message being stored. 
2. Every time I save a message as draft within a thread, we are creating a new draft message. It should ideally update the existing draft, or we should have a way of showing the draft message (like gmail).

yifanplanet commented on components/composer/src/init.spec.js:
```diff
@@ -852,3 +852,92 @@ describe("Composer subject", () => {
+      const component = el[0];
+      component.value = {
+        from: [
+          {
+            name: "Luka Test",
+            email: "luka.b@nylas.com",
+          },
+        ],
+        to: [
+          {
+            name: "Dan Test",
+            email: "dan.r@nylas.com",
+          },
```
Thanks for the review! To respond your change request, I've made a few changes in my new commit:
- I changed the button to be better semantically described by using interactive elements and to be tabbable.
- Added prop `show_save_as_draft` to show save draft button when only true, default to true.
- Added callback `afterSaveSuccess` and `afterSaveError`.
- The test cases used callback function for saving draft which doesn't test the actual API endpoints.

yifanplanet commented:
> @yifanplanet A couple things I noticed:
> 
> 1. When a message has an attachedFile and is saved as a draft, we should make an additional request to `/files` endpoint, get the file object and add it to the draft message being stored.
> 2. Every time I save a message as draft within a thread, we are creating a new draft message. It should ideally update the existing draft, or we should have a way of showing the draft message (like gmail).

@pooja169usp Thanks for the review! For your questions:
- I updated the code in my latest commit which update draft when message `id` and `version` is provided. For displaying and editing existing draft messages in mailbox, I've added the shortcut story specifically for that. [shortcut 81020](https://app.shortcut.com/nylas/story/81020/display-and-edit-draft-as-individual-message-in-a-email-component-thread)
- There's no need to add request to `/files` endpoints since it's already called when the file is attached and the file id is updated along with draft message and will be saved together as draft.
